lib/compress/zstd_compress.c

```diff
@@ -1820,34 +1820,41 @@ static size_t ZSTD_compressBlock_internal(ZSTD_CCtx* zc,
                                         const void* src, size_t srcSize)
 {
+    ZSTD_matchState_t* const ms = &zc->blockState.matchState;
     DEBUGLOG(5, "ZSTD_compressBlock_internal (dstCapacity=%u, dictLimit=%u, nextToUpdate=%u)",
-                (U32)dstCapacity, zc->blockState.matchState.dictLimit, zc->blockState.matchState.nextToUpdate);
+                (U32)dstCapacity, ms->dictLimit, ms->nextToUpdate);
     if (srcSize < MIN_CBLOCK_SIZE+ZSTD_blockHeaderSize+1)
         return 0;   /* don't even attempt compression below a certain srcSize */
     ZSTD_resetSeqStore(&(zc->seqStore));
 
     /* limited update after a very long match */
-    {   const BYTE* const base = zc->blockState.matchState.base;
+    {   const BYTE* const base = ms->base;
         const BYTE* const istart = (const BYTE*)src;
         const U32 current = (U32)(istart-base);
-        if (current > zc->blockState.matchState.nextToUpdate + 384)
-            zc->blockState.matchState.nextToUpdate = current - MIN(192, (U32)(current - zc->blockState.matchState.nextToUpdate - 384));
+        if (current > ms->nextToUpdate + 384)
+            ms->nextToUpdate = current - MIN(192, (U32)(current - ms->nextToUpdate - 384));
     }
 
     /* select and store sequences */
-    {   U32 const extDict = zc->blockState.matchState.lowLimit < zc->blockState.matchState.dictLimit;
+    {   U32 const extDict = ms->lowLimit < ms->dictLimit;
         size_t lastLLSize;
         { int i; for (i = 0; i < ZSTD_REP_NUM; ++i) zc->blockState.nextCBlock->rep[i] = zc->blockState.prevCBlock->rep[i]; }
         if (zc->appliedParams.ldmParams.enableLdm) {
-            typedef size_t (*ZSTD_ldmBlockCompressor)(
-                    ldmState_t* ldms, ZSTD_matchState_t* ms, seqStore_t* seqStore,
-                    U32 rep[ZSTD_REP_NUM], ZSTD_CCtx_params const* params,
-                    void const* src, size_t srcSize);
-            ZSTD_ldmBlockCompressor const ldmBlockCompressor = extDict ? ZSTD_compressBlock_ldm_extDict : ZSTD_compressBlock_ldm;
-            lastLLSize = ldmBlockCompressor(&zc->ldmState, &zc->blockState.matchState, &zc->seqStore, zc->blockState.nextCBlock->rep, &zc->appliedParams, src, srcSize);
+            size_t const nbStoredSeq =
+                ZSTD_ldm_generateSequences(&zc->ldmState, zc->ldmSequences,
+                                           ms, &zc->appliedParams.ldmParams,
+                                           src, srcSize, extDict);
+            size_t nbSeq = nbStoredSeq;
+            lastLLSize =
+                ZSTD_ldm_blockCompress(zc->ldmSequences, &nbSeq,
+                                       ms, &zc->seqStore,
+                                       zc->blockState.nextCBlock->rep,
+                                       &zc->appliedParams.cParams,
+                                       src, srcSize, extDict);
+            assert(nbSeq == nbStoredSeq /* Every stored sequence should be executed */);
```

Comment:
This check is surprising at this place.

While the documentation states that "`*nbSeq` is set to the number of sequences consumed", `ZSTD_ldm_blockCompress()` doesn't seem to modify this value anyway.

Is that related to "handle chunk sizes greater than the block size" ?

With current scope, it seems clearer to just pass `nbSeq` as a normal (read-copy) parameter, and make this check (if it's useful) inside `ZSTD_ldm_blockCompress()` instead. Thoughts ?

Reply:
Yeah, I can do that, and then revert it in the next PR.

The next thing I need to do on the LDM side is allow `zstdmt` to generate the LDM sequences for the whole chunk while holding a lock, and then compress blocks after that.